google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryWatchTest.java

```diff
@@ -337,15 +337,84 @@ public void nonEmptyResults_modifiedDocument_REMOVED() throws Exception {
     listenerAssertions.noError();
     listenerAssertions.eventCountIsAnyOf(Range.closed(2, 2));
     listenerAssertions.addedIdsIsAnyOf(emptyList(), singletonList("doc"));
-    listenerAssertions.modifiedIdsIsAnyOf(emptyList());
+    listenerAssertions.modifiedIdsIsEmpty();
     listenerAssertions.removedIdsIsAnyOf(emptyList(), singletonList("doc"));
 
-    ListenerEvent event = receivedEvents.get(receivedEvents.size() - 1);
+    ListenerEvent event = listener.lastListenerEvent();
     //noinspection ConstantConditions guarded by "assertNoError" above
     QueryDocumentSnapshot doc = event.value.getDocumentChanges().get(0).getDocument();
     assertThat(doc.get("foo")).isEqualTo("bar");
   }
 
+  /**
+   *
+   *
+   * <ol>
+   *   <li>Attach a listener to a query with non-empty results.
+   *   <li>Trigger existence filter mismatch, thereby invoking retry behavior.
+   *   <li>Add and remove documents.
+   *   <li>Verify expected snapshots are raised.
+   * </ol>
+   */
+  @Test
+  public void restartAfterFailedFilter() throws Exception {
+    // create a document in our collection that will match the query
+    DocumentReference testDoc1 = setDocument("doc1", map("foo", "bar"));
+    DocumentReference testDoc2 = setDocument("doc2", map("foo", "bar"));
+
+    final Query query = randomColl.whereEqualTo("foo", "bar");
+    QuerySnapshotEventListener listener =
+        QuerySnapshotEventListener.builder()
+            .setInitialEventCount(1)
+            .setAddedEventCount(3)
+            .setRemovedEventCount(1)
+            .build();
+    ListenerRegistration registration = query.addSnapshotListener(listener);
+
+    try {
+      listener.eventsCountDownLatch.awaitInitialEvents();
+      listener
+          .assertionsForLastEvent()
+          .noError()
```

Comment:
Add more comments explaining why we have this test and what would happen if SuppressiableStream is not used.